articles/react-error-handling.md

```diff
@@ -0,0 +1,436 @@
+
+```tsx
+<ErrorBoundary FallbackComponent={ErrorFallback} onReset={getProductData}>
+  {renderContent()}
+</ErrorBoundary>
+```
+
+The function getProductData is the result of refactoring the useEffect:
+
+```tsx
+const getProductData = useCallback(() => {
+  if (!productId) {
+    return
+  }
+
+  setStatus('pending')
+  makeRequest(productId)
+    .then((data) => {
+      setStatus('resolved')
+      setProductData(data)
+    })
+    .catch((error: Error) => {
+      setStatus('rejected')
+      setError(error.message)
+    })
+}, [productId])
+
+React.useEffect(() => {
+  getProductData()
+}, [getProductData])
+```
+
+We isolated the logic of making the request and updating the status in a function so we can re use it inside the useEffect and outside as we do passing it as prop. By the way, the function is wrapped in an useCallback to be only updated in case productId changed.
+
+It's time to test our application:
+
+![product-search-error-boundary-try-again.gif](/images/articles/react-error-handling/product-search-error-boundary-try-again.gif)
+
+It works fine, every time we got and error the Fallback component is shown and we could execute the reset function that makes the request once again then the components passed as chidren are re mounted.
+
+Let's analyze the following outcome: 
+
+![product-search-error-boundary-reset.gif](/images/articles/react-error-handling/product-search-error-boundary-reset.gif)
+
+Fallback component is shown but hitting the search button does not do anything looking at our button's handler we will notice we are taking the inputRef's value and setting it in the state productId. Well ErrorBoundary has an approach passing the prop `resetKeys`, its value will be a key (it could be our productId) and every time this key changes ErrorBoundary will reset its internal state. There is something we should keep in mind but let's figure it out by implement it and then analyze it:
+
+```tsx
+<ErrorBoundary 
+  FallbackComponent={ErrorFallback} 
+  onReset={getProductData}
+  resetKeys={[productId]} // ++
+>
+  {renderContent()}
+</ErrorBoundary>
+```
+
+Implementing this feature is quite easy we just need to pass a new prop. Let's see the outcome:
+
+![product-search-error-boundary-reset-key.gif](/images/articles/react-error-handling/product-search-error-boundary-reset-key.gif)
+
+Did you get it? The first error is thrown as we set it then hitting the search button does not work but if we change the input's value and hit the button it will work. Well, we need to review the search button's handler, it is taking the inputRef's value and then setting it as productId value. If you take a look back at the outcome we are hitting the search button with the same productId (the state is not changing) so the ErrorBoundary does not have to reset its state due to resetKeys prop has not changed. On the other hand when we change the input's value it counts as a change and ErrorBoundary will reset it.
```

Comment:
```suggestion
Did you get it? The first error is thrown as we set it, then hitting the search button does not work but if we change the input's value and hit the button it will work. Well, we need to review the search button's handler, it is taking the inputRef's value and then setting it as the productId value. If you take a look back at the outcome we are hitting the search button with the same productId (the state is not changing) so the ErrorBoundary does not have to reset its state due to resetKeys prop has not changed. On the other hand when we change the input's value it counts as a change and ErrorBoundary will reset it.
```